Jenkins.io - Create platform information section (#6685)
jenkins-infra/jenkins.io: content/doc/book/platform-information/upgrade-java-to-11.adoc

---
layout: subsection
title: Upgrading to Java 11
---

There are a few details and steps to upgrade the JVM used to run Jenkins, more specifically from Java 8 to Java 11.

If you're upgrading the JVM used to run Jenkins, and particularly if you're upgrading from Java 8 to Java 11, there are some details you should know and precautions you should take.

video::L2Uomz8RWUM[youtube,width=800,height=420]

> == Backup

Reviewer: ```suggestion
== Backing up Jenkins
```
This is what is used in the other section.

> As with any upgrade, we recommend backing up `JENKINS_HOME` and testing the upgrade with the backup before performing the upgrade on your production instance.

Reviewer: I would include the link to backing up `JENKINS_HOME` in the steps below.

> As with any upgrade, we recommend backing up `JENKINS_HOME` and testing the upgrade with the backup before performing the upgrade on your production instance.

Reviewer: I would suggest using a numbered list to make the recommended steps very clear.

// Commented because pipeline support plugin 3.0 is over 3 years old and has 8+ later releases
//
// One of the most important plugin upgrades is the plugin:workflow-support[Pipeline: Support plugin]: make sure that the version of the plugin is at least `3.0`.
//
// NOTE: Stop all Pipeline jobs before upgrading this plugin because this upgrade changes the serialization of Pipeline builds. As a general rule, even though Pipeline jobs are supposed to survive a Jenkins restart, it's always a better option to make sure that no Pipeline builds are in progress before any scheduled Jenkins maintenance.

== Jakarta XML Binding

Some plugins use JAXB libraries provided by the JDK.
However, the `java.xml.bind` and `javax.activation` modules are no longer included in OpenJDK 11, and plugins might fail if no replacement is offered.

To fix this problem, we've bundled those libraries into a new detached plugin: plugin:jaxb[JAXB plugin].
When any Jenkins core more recent than `2.163` is running on Java 11, this plugin is automatically installed.

> However, if you manage your plugins outside Jenkins, for example, if you use `plugins.txt` in your Docker images, you might need to install the plugin explicitly.

Reviewer: ```suggestion
However, if you manage your plugins outside Jenkins, for example using a `plugins.txt` in your Docker images, you might need to install the plugin explicitly.
```
Just helps cut down on commas.

== JVM version on agents

All agents must be running on the same JVM version as the controller due to how controllers and agents communicate.

> If you're upgrading your Jenkins controller to run on Java 11, you also need to upgrade the JVM on your agents.

Reviewer: ```suggestion
If you're upgrading your Jenkins controller to run on Java 11, you must upgrade the JVM on your agents.
```

You can validate the version of each agent with the plugin:versioncolumn[Versions Node Monitors] plugin.
This plugin provides information about the JVM version of each agent on the node management screen of your Jenkins instance.
You can also configure this plugin to automatically disconnect any agent with an incorrect JVM version.

== Java Web Start

Java Web Start has been removed in Java 11.
When a Jenkins controller runs on Java 11, the Java Web Start button will no longer appear in the Web UI.
You can't launch agents for a Java 11 Jenkins controller from a `*.jnlp` file downloaded to a web browser.

There are no plans to replace this functionality.
Connect agents to Jenkins on Java 11 with plugins like plugin:ssh-slaves[SSH Build Agents Plugin], with operating system command line calls to `java -jar agent.jar`, or using containers.

> == JDK Tool Automatic installer

Reviewer: ```suggestion
== JDK Tool Automatic installer
```
Not sure if the capitalization here is correct. I couldn't find a reference to it.

Author: I found a reference, it does seem to be the name of the tool https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/java-tool-installer-v0?view=azure-pipelines
